lingpy/sequence/ngrams.py

```diff
@@ -0,0 +1,685 @@
+    if not single_gap_opening:
+        # We pad the `sequence` with None symbols to the right,
+        # so we can filter during the list comprehension.
+        # Please note that this is *not* the user-requested
+        # padding, but an internal and inexpansive way to
+        # account for the end of the ngram; also
+        # please note that this will fail if the sequence itself
+        # holds None symbols.
+        # NOTE: To solve the problems when/if the sequence itself
+        #       holds None symbols, we could translate Nones to
+        #       a tempory value and remap it when yielding; while
+        #       expansive, this is still more effective than
+        #       other solutions which would not allow using
+        #       `all()` for most sequence computations.
+        _temp = chain(seq, (None,) * order)
+        for ngram in get_n_ngrams(_temp, order+max_gaps, pad_symbol=None):
+            head = ngram[:1] # cache for all combinations
+            for comb in [tail for tail in combinations(ngram[1:], order-1) if all(tail)]:
+                yield head + comb
+    else:
+        # Iterate over all the possible gap lengths, including
+        # length zero. Length zero requires a different logic (it is
+        # actually just returning all subsequent ngrams) in order
+        # not to yield the same subsequence (for example, for length
+        # 1+2 and 2+1, which are obviously identical). One alternative,
+        # would be to add a function paramenter allowing the user to
+        # circumvent this behaviour, but no situation where this would
+        # be required or at least suggested seem to exist (we would be
+        # distributing more probability to ngrams with no gaps, which
+        # is exactly what skip ngrams are intended to correct).
+        for gap_width in range(max_gaps+1):
+            if gap_width == 0:
+                # No pad-symbol here, as the sequence was padded before.
+                for ngram in get_n_ngrams(seq, order, pad_symbol=None):
+                    yield ngram
+            else:
+                # We iterate over all possible left and right lengths,
+                # making sure we always have at least one on each side.
+                for left_width in range(1, order):
+                    # Build the pattern we are extracting
+                    pattern = (True,) * left_width + \
+                              (False,) * gap_width + \
+                              (True,) * (order - left_width) # right width
+
+                    # Iterate over the sequence while applying the pattern;
+                    # `idx` is the starting index in `sequence`, `skip` the
+                    # element index in `pattern`
+                    for idx in range(len_seq-order-gap_width+1):
+                        yield tuple(seq[idx+skip] for skip, keep in enumerate(pattern) if keep)
+
+def get_posngrams(sequence, pre_order=0, post_order=0, pad_symbol='$', elm_symbol='###'):
```

Comment:
`get_pos_ngrams`

Reply:
See above, this would be actually confusing, suggesting that (traditional) ngrams and positional ngrams are the same. But we could go for `get_p_posngrams` or `get_all_posngrams`.